Add V4 distribution implementation developer guide (#4783)
pymc-devs/pymc: docs/source/distribution_implementation.md

# Implementing a Distribution (developer guide)

This guide provides an overview on how to implement a distribution for version 4 of PyMC3.
It is designed for developers who wish to add a new distribution to the library.

> Users will not be aware of all this complexity and should instead make use of helper methods such as (TODO).

Reviewer: Leftover TODO

Author: It's still a pymc3 ToDo. This was the role of `Densitydist` but that was deprecated or at least not refactored for V4. Maybe we can link to something like: https://github.com/pymc-devs/pymc-examples/pull/185

# Implementing a Distribution (devs)

This guide provides an overview on how to implement a distribution for version 4
of PyMC3. It is designed for developers who wish to add a new distribution to
the library. Users will not be aware of all this complexity and should instead
make use of helper methods such as (TODO?)

PyMC3 distributions build on top of Aesara's random variables (TODO: add link),
and implement `logp` and `logcdf` methods as well as other initialization and
validation helpers, most notably `shape/dims`, alternative parametrizations, and
default `transforms`.

Here is a summary check-list of the steps needed to implement a new distribution.
Each section will be expanded below:

> 1. Creating a new `RandomVariable` `op`

Reviewer 2: general question, are there docs for RandomVariable and Distribution base classes and if so, do we want to link there?

Author: We should. We have docstrings for the distribution base classes. For the RandomVariable I am not sure if it appears in the aesara docs, but if not we can add something for sure

Reviewer: I find https://aesara.readthedocs.io/en/latest/extending/graphstructures.html to be one of the most useful chapters in the Aesara docs.

Reviewer: ```suggestion
1. Creating a new `RandomVariable` `Op`
```

1. Implementing the corresponding `Distribution` class
1. Adding tests for the new `RandomVariable`
1. Adding tests for the `logp` / `logcdf` methods
1. Documenting the new `Distribution`.

This guide does not attempt to explain the rationale behind the `Distributions`
current implementation, and details are provided only insofar as they help

> implementing new "standard" distributions.

Reviewer 3: >  This guide does not attempt to explain the rationale behind the `Distributions` current implementation

Would you say that [this PR](https://github.com/pymc-devs/pymc3/pull/4440) is a good reference for anyone interested in that? if so, would it make sense to add it here for anyone interested in better understanding the ratio behind the change in implementation?

Author: After having a look, I feel it's too unstructured to be of real help. Besides it was not even merged in the end. This one contains all the changes, but it's gigantic: https://github.com/pymc-devs/pymc3/pull/4500

## 1. Creating a new `RandomVariable` `op`

`RandomVariables` are responsible for implementing the random sampling methods,
which in version 3 of PyMC3 used to be one of the standard `Distribution`
methods, alongside `logp` and `logcdf`. The `RandomVariable` is also responsible
for parameter broadcasting and shape inference.

Before creating a new `RandomVariable` make sure that it is not offered in the

> [Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions)

Reviewer 2: I think this could be `` {class}`numpy.random.Generator` `` or `` {class}`Numpy library <numpy.random.generator>` `` provided we set up a numpy intershpinx mapping, which will probably come in handy in other places,

Reviewer: ```suggestion
[Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions).
```

   and should not make use of other `Aesara` symbolic ops. The random method
   should make use of the `rng` which is a Numpy `Random State / Generator` ??,
   so that samples are reproducible.
1. The `size` argument (together with the inputs shape) is the only way for the
   user to specify the (shape?, number of draws?) of the `RandomVariable`.
   The `rng_fn` will have to take this into consideration for correct output.
   `size` is the specification used by `Numpy` and `Scipy` and works like
   PyMC3 `shape` for univariate distributions, but is different for multivariate
   distributions. Unfortunately there is no general reference documenting how
   `size` ought to work for multivariate distributions.
   This [discussion](https://github.com/numpy/numpy/issues/17669) may be helpful
   to get more context.
1. `Aesara` tries to infer the output shape of the `RandomVariable`
   (given a user-specified size) by introspection of the `ndim_supp` and
   `ndim_params` attributes. However, the default method may not work for more
   complex distributions. In that case, custom `_shape_from_params` (and less
   probably (?) `_infer_shape`) should also be implemented in the new

>    `RandomVariable` class. One simple example is seen in the [MultinomialRV](https://github.com/aesara-devs/aesara/blob/3a3adaee3b8ceccff6d0208996f117fcd5e5eb6f/aesara/tensor/random/basic.py#L518-L521)

Reviewer 2: It looks like MultinomialRV is not anywhere within Aesara's docs, once it is I think this would be better as an intersphinx link. What I don't know how it can be done is linking directly to the source code instead of link to api page that then has the `[source]` link :thinking:

Author: The Dirichlet multinomial will have the same trick, so we can link to that #4758

   where it was necessary to specify the `rep_param_idx` so that the
   `default_shape_from_params` helper method could do its job. In more complex
   cases, it may not be possible to make use of the default helper,
   but those have not been found yet!
1. It's okay to use the `rng_fn` `classmethods` of other Aesara and PyMC3
   `RandomVariables` inside the new `rng_fn`. For example if you are
   implementing a negative HalfNormal `RandomVariable`, your `rng_fn` can simply
   return `- halfnormal.rng_fn(rng, scale, size)`.

For a quick test that your new `RandomVariable` `Op` is working, you can just
call `eval` on a new instance:

> ```python

Reviewer 2: you have structured all the code as input and output with outputs within the python highlighted code cell, would it make sense to make the file executable? and either run it every time we build docs or save it as ipynb so output gets stored and we don't need to run every time?

Author: I am not sure how functional the pseudo code is :/

# blah = aesara.tensor.random.uniform in this example
blah([0, 0], [1, 2], size=(10, 2)).eval()

# array([[0.83674527, 0.76593773],
#    [0.00958496, 1.85742402],
#    [0.74001876, 0.6515534 ],
#    [0.95134629, 1.23564938],
#    [0.41460156, 0.33241175],
#    [0.66707807, 1.62134924],
#    [0.20748312, 0.45307477],
#    [0.65506507, 0.47713784],
#    [0.61284429, 0.49720329],
#    [0.69325978, 0.96272673]])

```

## 2. Inheriting from a PyMC3 base `Distribution` class

After implementing the new `RandomVariable` `op`, it's time to make use of it

> in a new PyMC3 `distribution`. PyMC3 works in a very functional way, and the

Author: We can try that glossary thing with an entry for `functional programming`

         return bound(
            logp_expression,
            value >= 0,
            param2 >= 0,
            # There is one sneaky optional keyowrd argument, that converts an
            # otherwise elemwise `bound` to a reduced scalar bound. This is usually
            # needed for multivariate distributions where the dimensionality
            # of the bound conditions does not match that of the "value" / "logp"
            # By default it is set to `True`.
            broadcast_conditions=True,
        )

    # logcdf works the same way as logp. For bounded variables, it is expected
    # to return `-inf` for values below the domain start and `0` for values
    # above the domain end, but only when the parameters are valid.
    def logcdf(value, param1, param2):
        ...

```

Some notes:

1. A distribution should at the very least inherit from `Discrete` or
   `Continuous`. For the latter, more specific subclasses exist:
   `PositiveContinuous`, `UnitContinuous`, `BoundedContinuous`,
   `CircularContinuous`, which specify default transformations for the variables.
   If you need to specify a one-time custom transform you can also override the

>    `__new__` method, as is done for the [`Dirichlet`](https://github.com/pymc-devs/pymc3/blob/6eaf9a0d7bc5b9168bedc9337808391ccb86a5ea/pymc3/distributions/multivariate.py#L399-L401).

Reviewer 2: This could also be `` {class}`pymc3.Dirichlet` ``, but again, this goes to the api docs that will have the `[source]` link so it's two clicks, I'll try to find a better way. 

I don't like references to github lines because they either point to a specific commit and get outdated without anyone realizing so checking the link goes to the right code but that code is no longer in use within the actual library or they point to the main/dev branch and any change in line numbers makes the link point to the right file but wrong section.

1. If a distribution does not have a corresponding `random` implementation, a
   `RandomVariable` should still be created that raises a `NotImplementedError`.

>    This is the case for the [`Flat` distribution](https://github.com/pymc-devs/pymc3/blob/6eaf9a0d7bc5b9168bedc9337808391ccb86a5ea/pymc3/distributions/continuous.py#L312-L321).

Reviewer 2: With FlatRV the situation is similar yet a bit different, because FlatRV _in not part of the docs!_. I understand it should not be part of the user facing docs but it looks like it's useful for dev facing docs. Would it be a good idea to have a couple of orphan API pages (therefore not indexed in the API navbar section) so we can link there from the developer guide?

Author: Sounds reasonable

   run in a pure "float64" environment, which means the reference and the `Pymc3`
   results can diverge quite a lot (e.g., underflowing to `-np.inf` for extreme
   parameters). You should therefore make sure you test locally in both regimes.
   A quick and dirty way of doing this is to temporariliy add
   `aesara.config.floatX = "float32"` at the very top of file, immediately after
   `import aesara`. Remember to set `n_samples=-1` as well to test all
   combinations. The test output will show what exact parameter values lead to
   a failure. If you are confident that your implementation is correct, you may
   opt to tweak the decimal precision with `select_by_precision`, or adjust the
   tested `Domain` values. In extreme cases, you can mark the test with a
   conditional `xfail` (if only one of the sub-methods is failing, they should
   be separated, so that the `xfail` is as narrow as possible):

```python

test_blah_logp(self):
    ...

@pytest.mark.xfail(
   condition=(aesara.config.floatX == "float32"),
   reason="Fails on float32 due to numerical issues",
)
test_blah_logcdf(self):
    ...

```

## 5. Documenting the new `Distribution`

> TODO: What needs to be done? Adding a new entry to the .rst?

Reviewer 2: At the very least, they should have a docstring written and then be referenced in the respective API page. Depending on the distribution and it's uses it should probably get a notebook example where it's used in a case study or a tutorial if it has significant difference all other distributions.

# Implementing a Distribution (developer guide)

This guide provides an overview on how to implement a distribution for version 4
of PyMC3. It is designed for developers who wish to add a new distribution to
the library. Users will not be aware of all this complexity and should instead

> make use of helper methods such as (TODO).

Reviewer: To reduce diff complexity with markdown files, one can do one line per sentence. Single line breaks don't show up in the rendered output and with text I don't think we need to have a limit on the characters per line?

Author: Done

PyMC3 {`Distributions`}`pymc3.distributions.Distribution` build on top of

> Aesara's {`RandomVariables`}`aesara.tensor.random.op.RandomVariable`, and

Reviewer 2: I now realize I did not share any reference on how intersphinx works. The `{class}` thingy indicates that we are linking to an object of type class. intersphinx docs are quite bad so I tried to document everything and add examples at https://arviz-devs.github.io/arviz/contributing/developer_guide.html#reference-external-libraries. It is still missing the custom text option though:

```
{class}`pymc3.distributions.Distribution`                 -> pymc3.distributions.Distribution
{class}`~pymc3.distribution.Distribution`                 -> Distribution
{class}`Distributions <pymc3.distribution.Distribution>`  -> Distributions
```

implement `logp` and `logcdf` methods as well as other initialization and
validation helpers, most notably `shape/dims`, alternative parametrizations, and
default `transforms`.

Here is a summary check-list of the steps needed to implement a new distribution.
Each section will be expanded below:

1. Creating a new `RandomVariable` `op`
1. Implementing the corresponding `Distribution` class
1. Adding tests for the new `RandomVariable`
1. Adding tests for the `logp` / `logcdf` methods
1. Documenting the new `Distribution`.

This guide does not attempt to explain the rationale behind the `Distributions`
current implementation, and details are provided only insofar as they help
to implement new "standard" distributions.

> ## 1. Creating a new `RandomVariable` `op`

Reviewer: ```suggestion
## 1. Creating a new `RandomVariable` `Op`
```

implementing the random sampling methods, which in version 3 of PyMC3 used to be
one of the standard `Distribution` methods, alongside `logp` and `logcdf`. The
`RandomVariable` is also responsible for parameter broadcasting and shape
inference.

Before creating a new `RandomVariable` make sure that it is not offered in the
[Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions)
If it is, it should be added to the [Aesara library](https://github.com/aesara-devs/aesara)
first and then imported into the PyMC3 library.

In addition, it might not always be necessary to implement a new `RandomVariable`.
For example if the new `Distribution` is just a special parametrization of an
existing `Distribution`. This is the case of the `OrderedLogistic` and
`OrderedProbit`, which are just special parametrizations of the `Categorial`
distribution.

The following snippet illustrates how to create a new `RandomVariable`:

```python

from aesara.tensor.var import TensorVariable
from aesara.tensor.random.op import RandomVariable

# Create your own `RandomVariable`...
class BlahRV(RandomVariable):
    name: str = "blah"

>     # Provide the number of (output) dimensions for this RV

Reviewer: ```suggestion
    # Provide the minimum number of (output) dimensions for this RV
```

    # (e.g. `0` for a scalar, `1` for a vector, etc.)
    ndim_supp: int = 0

    # Provide the number of (input) dimensions for each parameter of the RV
    # (e.g. if there's only one vector parameter, `[1]`; for two parameters,
    # one a matrix and the other a scalar, `[2, 0]`; etc.)
    ndims_params: List[int] = [0, 0]

    # The NumPy/Aesara dtype for this RV (e.g. `"int32"`, `"int64"`).
    # The standard in the library is `"int64"` for discrete variables
    # and `"floatX"` for continuous variables
    dtype: str = "floatX"

    # A pretty text and LaTeX representation for the RV

>     _print_name: Tuple[str, str] = ("blah", "\\operatorname{blah}")

Reviewer: Maybe choose a more meaningful name so it becomes clear how capitalization should be for these Text/LaTeX fields?

`MoonshotRV` or something.

Author: What would that show differently, underscores?

Reviewer: I'm asking because the LaTeX operators for `Uniform`, or `Normal` start with capital letters, right? But this example does not...

Author: Ahhh... I'll change it then. Missed that one

Author: ```suggestion
    _print_name: Tuple[str, str] = ("Blah", "\\operatorname{Blah}")
```

    # If you want to add a custom signature and default values for the
    # parameters, do it like this. Otherwise this can be left out.
    def __call__(self, loc=0.0, scale=1.0, size=None, **kwargs) -> TensorVariable:
        return super().__call__(loc, scale, size=size, **kwargs)

    # This is the Python code that produces samples.  Its signature will always
    # start with a NumPy `RandomState` object, then the distribution
    # parameters, and, finally, the size.
    #
    # This is effectively the v4 replacement for `Distribution.random`.
    @classmethod
    def rng_fn(
        cls,
        rng: np.random.RandomState,
        loc: np.ndarray,
        scale: np.ndarray,
        size: Tuple[int, ...],
    ) -> np.ndarray:
        return scipy.stats.blah.rvs(loc, scale, random_state=rng, size=size)

# Create an instance of the `RandomVariable` `Op`...

> blah = BlahRV()

Reviewer: The user API is always through things like `pm.Normal`, so what's the purpose of these instances?

Author: This is what actually creates the `RandomVariable Op`.

Author: Added a more clear comment, not sure if I am saying something incorrect

```

Some important things to keep in mind:

1. Everything inside the `rng_fn` method is pure Python code (as are the inputs)
   and should not make use of other `Aesara` symbolic ops. The random method
   should make use of the `rng` which is a Numpy
   {`Random State`}`numpy.random.RandomState`, so that samples are reproducible.

> 1. The `size` argument (together with the inputs shape) is the only way for the
>    user to specify the `batch_shape` (~number of repetitions) of the
>    `RandomVariable`.
>    The `rng_fn` will have to take this into consideration for correct output.
>    `size` is the specification used by `Numpy` and `Scipy` and works like
>    PyMC3 `shape` for univariate distributions, but is different for multivariate
>    distributions. Unfortunately there is no general reference documenting how
>    `size` ought to work for multivariate distributions.
>    This [discussion](https://github.com/numpy/numpy/issues/17669) may be helpful
>    to get more context.

Reviewer: ```suggestion
1. The `size` argument (together with the inputs shape) are the one way to specify
    the dimensions (minus the last `ndim_supp`) at the `RandomVariable` API level.
    At the user API level more options (`shape`, `dims`) are available, but this is taken
    care of by {`class`}`~pymc3.distributions.Distribution`.
   The `rng_fn` will have to take `size` into consideration for correct output.
   `size` is the specification used by `Numpy` and `Scipy` and works like
   PyMC3 `shape` for univariate distributions, but is different for multivariate
   distributions. Unfortunately there is no general reference documenting how
   `size` ought to work for multivariate distributions.
   This [discussion](https://github.com/numpy/numpy/issues/17669) may be helpful
   to get more context.
```

Author: Added a slightly different note on this below.

1. `Aesara` tries to infer the output shape of the `RandomVariable`
   (given a user-specified size) by introspection of the `ndim_supp` and
   `ndim_params` attributes. However, the default method may not work for more
   complex distributions. In that case, custom `_shape_from_params` (and less
   probably, `_infer_shape`) should also be implemented in the new
   `RandomVariable` class. One simple example is seen in the
   {MultinomialRV}`pymc3.distributions.multivariate.DirichletMultinomialRV`
   where it was necessary to specify the `rep_param_idx` so that the
   `default_shape_from_params` helper method could do its job. In more complex
   cases, it may not be possible to make use of the default helper,
   but those have not been found yet!
1. It's okay to use the `rng_fn` `classmethods` of other Aesara and PyMC3
   `RandomVariables` inside the new `rng_fn`. For example if you are
   implementing a negative HalfNormal `RandomVariable`, your `rng_fn` can simply
   return `- halfnormal.rng_fn(rng, scale, size)`.

> For a quick test that your new `RandomVariable` `Op` is working, you can just
> call `eval` on a new instance:

Reviewer: ```suggestion
For a quick test that your new `RandomVariable` `Op` is working, you can call
the `Op` with parameters and call `eval()` on the resulting {class}`~aesara.tensor.var.TensorVariable` instance:
```

```python

# blah = aesara.tensor.random.uniform in this example
blah([0, 0], [1, 2], size=(10, 2)).eval()

# array([[0.83674527, 0.76593773],
#    [0.00958496, 1.85742402],
#    [0.74001876, 0.6515534 ],
#    [0.95134629, 1.23564938],
#    [0.41460156, 0.33241175],
#    [0.66707807, 1.62134924],
#    [0.20748312, 0.45307477],
#    [0.65506507, 0.47713784],
#    [0.61284429, 0.49720329],
#    [0.69325978, 0.96272673]])

```

## 2. Inheriting from a PyMC3 base `Distribution` class

> After implementing the new `RandomVariable` `op`, it's time to make use of it

Reviewer: ```suggestion
After implementing the new `RandomVariable` `Op`, it's time to make use of it
```

in a new PyMC3 {`Distribution`}`pymc3.distributions.Distribution`. PyMC3 works
in a very {functional}`Functional_Programming` way, and the `distribution`

> classes are there mostly to facilitate porting the `v3` code to the new `v4`
> version, and to keep related methods organized together. In practice, they take
> care of:

Reviewer: ```suggestion
classes are there mostly to facilitate porting the `v3` code to the new `v4`
version, add PyMC3 API features and keep related methods organized together.
In practice, they take
care of:
```

        )

    # logcdf works the same way as logp. For bounded variables, it is expected
    # to return `-inf` for values below the domain start and `0` for values
    # above the domain end, but only when the parameters are valid.
    def logcdf(value, param1, param2):
        ...

```

Some notes:

1. A distribution should at the very least inherit from `Discrete` or
   `Continuous`. For the latter, more specific subclasses exist:
   `PositiveContinuous`, `UnitContinuous`, `BoundedContinuous`,
   `CircularContinuous`, which specify default transformations for the variables.
   If you need to specify a one-time custom transform you can also override the
   `__new__` method, as is done for the
   {`Dirichlet`}`pymc3.distributions.multivariate.Dirichlet`.
1. If a distribution does not have a corresponding `random` implementation, a
   `RandomVariable` should still be created that raises a `NotImplementedError`.
   This is the case for the {`Flat`}`pymc3.distributions.continuous.FlatRV`.
   In this case it will be necessary to provide a standard `initval` by
   overriding `__new__`.
1. As mentioned above, `v4` works in a very {functional}`Functional_Programming`

>    way, and all the   information that is needed in the `logp` and `logcdf`

Reviewer: ```suggestion
   way, and all the information that is needed in the `logp` and `logcdf`
```

   methods is expected to be "carried" via the `RandomVariable` inputs. You may

>    pass numerical arguments that are not strictly needed for the `random` method

Reviewer: ```suggestion
   pass numerical arguments that are not strictly needed for the `rng_fn` method
```
?

for the associated default parameter. For instance if it's just the inverse,
testing with `1.0` is not very informative, since the conversion would return
`1.0` as well, and we can't be (as) sure that is working correctly.

```python

class TestBlahAltParam2(BaseTestDistribution):

    pymc_dist = pm.Blah
    # param2 is equivalent to 1 / alt_param2
    pymc_dist_params = {"param1": 0.25, "alt_param2": 4.0}
    expected_rv_op_params = {"param1": 0.25, "param2": 2.0}
    tests_to_run = ["check_pymc_params_match_rv_op"]

```

Custom tests can also be added to the class as is done for the
{`FlatRV`}`pymc3.tests.test_random.TestFlat`.

### Note on `check_rv_size` test:

Custom input sizes (and expected output shapes) can be defined for the
`check_rv_size` test, by adding the optional class attributes `sizes_to_check`
and `sizes_expected`:

```python
sizes_to_check = [None, (1), (2, 3)]
sizes_expected = [(3,), (1, 3), (2, 3, 3)]
tests_to_run = ["check_rv_size"]
```

This is usually needed for Multivariate distributions. You can see an example in
the {`Dirichlet` test}`pymc3.test.test_random.TestDirichlet`.

> ### Notes on `check_pymcs_draws_match_reference` test

Reviewer: Maybe move this into a docstring right at the test definition?

   (they are used for other things). If it is important to test the edge of the
   `Domain`, the edge values can be repeated. This is done by the `Bool`:
   `Bool = Domain([0, 0, 1, 1], "int64")`
1. There are some default domains (such as `R` and `Rplus`) that you can use for
   testing your new distribution, but it's also perfectly fine to create your
   own domains inside the test function if there is a good reason for it (e.g.,
   when the default values lead too many extreme unlikely combinations that are
   not very informative about the correctness of the implementation).
1. By default, a random subset of 100 `param` * `paramdomain` combinations is
   tested, in order to keep the test runtime under control. When testing your
   shiny new distribution, you can temporarily set `n_samples=-1` to force all
   combinations to be tested. This is important to avoid the your `PR` leading
   to surprising failures in future runs whenever some bad combinations of
   parameters are randomly tested.
1. On Github all tests are run twice, under the `aesara.config.floatX` flags of
   `"float64"` and `"float32"`. However, the reference Python functions will

>    run in a pure "float64" environment, which means the reference and the `Pymc3`

Reviewer: ```suggestion
   run in a pure "float64" environment, which means the reference and the `PyMC3`
```

   combinations. The test output will show what exact parameter values lead to
   a failure. If you are confident that your implementation is correct, you may
   opt to tweak the decimal precision with `select_by_precision`, or adjust the
   tested `Domain` values. In extreme cases, you can mark the test with a
   conditional `xfail` (if only one of the sub-methods is failing, they should
   be separated, so that the `xfail` is as narrow as possible):

```python

def test_blah_logp(self):
    ...

@pytest.mark.xfail(
   condition=(aesara.config.floatX == "float32"),
   reason="Fails on float32 due to numerical issues",
)
def test_blah_logcdf(self):
    ...

```

## 5. Documenting the new `Distribution`

New distributions should have a rich docstring, following the same format as
that of previously implemented distributions. It generally looks something like
this:

```python
 r"""
 Univariate blah distribution.

 The pdf of this distribution is

 .. math::

    f(x \mid \param1, \param2) = \exp{x * (param1 + \log{param2})}

 .. plot::

     import matplotlib.pyplot as plt
     import numpy as np
     import scipy.stats as st
     import arviz as az

>      plt.style.use('arviz-darkgrid')

Reviewer: ```suggestion
```
These style changes are unrelated to the code that's being demonstrated. If we consider this important for the documentation we should set the style at docs-build-time.
